firestore/client.go

```diff
@@ -127,12 +159,25 @@ func (c *Client) path() string {
 }
 
 func withResourceHeader(ctx context.Context, resource string) context.Context {
-	md, _ := metadata.FromOutgoingContext(ctx)
+	md, ok := metadata.FromOutgoingContext(ctx)
+	if !ok {
+		md = metadata.New(nil)
+	}
 	md = md.Copy()
 	md[resourcePrefixHeader] = []string{resource}
 	return metadata.NewOutgoingContext(ctx, md)
 }
 
+func withRequestParamsHeader(ctx context.Context, requestParams string) context.Context {
```

Comment:
Nit: Not sure if this module already uses `gax-go` or not, but we have a helper to do this provided key-value pairs: https://pkg.go.dev/github.com/googleapis/gax-go/v2#InsertMetadataIntoOutgoingContext

```go
gax.InsertMetadataIntoOutgoingContext(ctx, reqParamsHeader, requestParams)
```

Feel free to keep your own code here, but might be good to dedupe it (or later).